inductor(cpu): support mkldnn packed linear to improve bfloat16 performance

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96954


As title, enable mkldnn packed linear to improve bfloat16 performance.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor(cpu): support mkldnn packed linear to improve bfloat16 performance

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor(cpu): support mkldnn packed linear to improve bfloat16 performance"




cc jgong5 mingfeima sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor(cpu): support mkldnn packed linear to improve bfloat16 performance"



As title, enable mkldnn packed linear to improve bfloat16 performance.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on torch/_inductor/mkldnn.py:
```diff
@@ -283,52 +283,75 @@ def forward(self, input):
-        linear: nn.Module,
-        unary: nn.Module,
-    ):
+    def __init__(self, linear: nn.Module, unary: Optional[nn.Module], input_size: list):
```
`unary` is optional now. So does it intend to support a single `Linear`?

XiaobingSuper replied:
Yes, you are right.

XiaobingSuper committed:
Update on "inductor(cpu): support mkldnn packed linear to improve bfloat16 performance"



As title, enable mkldnn packed linear to improve bfloat16 performance.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@jansel @desertfire , please help review this PR, thanks!

desertfire commented on test/inductor/test_torchinductor.py:
```diff
@@ -1945,12 +1945,13 @@ def forward(self, x):
+                m_opt = torch.compile(m)
+                x = torch.randn(224, 224, dtype=torch.bfloat16)
+                m_opt(x)
+                self.assertEqual(m(x), m_opt(x))
```
Nit: is there a way to check if the weight change did happen?

jgong5 replied:
I guess we have to let users to specify the weights are constant from the frontend, something like `torch.jit.freeze`.

XiaobingSuper replied:
Yes, Let's do it at the next step.

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper commented:
@pytorchbot rebase

XiaobingSuper committed:
Update on "inductor(cpu): support mkldnn packed linear to improve bfloat16 performance"



As title, enable mkldnn packed linear to improve bfloat16 performance.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
